lib/nostrum/api.ex

```diff
@@ -1355,54 +1360,82 @@ defmodule Nostrum.Api do
+  If the request is successful, this function returns `{:ok, user}`, where 
+  `user` is a `Nostrum.Struct.User`. Otherwise, returns `{:error, reason}`.
   """
-  @spec get_user(integer) :: error | {:ok, Nostrum.Sturct.User.t}
+  @spec get_user(User.id) :: error | {:ok, User.t}
   def get_user(user_id) do
     request(:get, Constants.user(user_id))
+    |> handle_request_with_decode({:struct, User})
   end
 
-  @spec get_user!(integer) :: no_return | Nostrum.Struct.User.t
+  @doc """
+  Same as `get_user/1`, but raises `Nostrum.Error.ApiError` in case of failure.
+  """
+  @spec get_user!(User.id) :: no_return | User.t
   def get_user!(user_id) do
     get_user(user_id)
     |> bangify
   end
 
   @doc """
   Gets info on the current user.
+
+  If nostrum's caching is enabled, it is recommended to use `Nostrum.Cache.Me.get/0` 
+  instead of this function. This is because sending out an API request is much slower 
+  than pulling from our cache.
+
+  If the request is successful, this function returns `{:ok, user}`, where 
+  `user` is nostrum's `Nostrum.Struct.User`. Otherwise, returns `{:error, reason}`.
   """
-  @spec get_current_user() :: error | {:ok, Nostrum.Struct.User.t}
+  @spec get_current_user() :: error | {:ok, User.t}
   def get_current_user do
-    case request(:get, Constants.me) do
-      {:ok, body} ->
-        {:ok, Poison.decode!(body, as: %User{})}
-      other ->
-        other
-    end
+    request(:get, Constants.me)
+    |> handle_request_with_decode({:struct, User})
+  end
+
+  @doc """
+  Same as `get_current_user/0`, but raises `Nostrum.Error.ApiError` in case of failure.
+  """
+  @spec get_current_user!() :: no_return | User.t
+  def get_current_user! do
+    get_current_user()
+    |> bangify
   end
 
   @doc ~S"""
   Changes the username or avatar of the current user.
 
-  **Example**
-  ```Elixir
-  avatar = %{avatar: "data:image/jpeg;base64," <> "YXl5IGJieSB1IGx1a2luIDQgc3VtIGZ1az8="}
-  {:ok, user} = Nostrum.Api.modify_current_user(avatar)
-  ```
+  ## Params 
+ 
+  The following params are optional: 
+ 
+    * `:username` (string) - new username
+    * `:avatar` (string) - the user's avatar as [avatar data](https://discordapp.com/developers/docs/resources/user#avatar-data)
+ 
+  ## Examples 
 
-  `options` is a map with the following optional keys:
-   * `username` - New username.
-   * `avatar` - Base64 encoded image data, prepended with `data:image/jpeg;base64,`
+      iex> Nostrum.Api.modify_current_user(avatar: "data:image/jpeg;base64,YXl5IGJieSB1IGx1a2luIDQgc3VtIGZ1az8=") 
+      {:ok, %Nostrum.Struct.User{}}
   """
-  def modify_current_user(options) do
-    case request(:patch, Constants.me, options) do
-      {:ok, body} ->
-        {:ok, Poison.decode!(body)}
-      other ->
-        other
-    end
+  @spec modify_current_user(keyword | map) :: error | {:ok, User.t}
+  def modify_current_user(params)
+  def modify_current_user(params) when is_list(params), do: modify_current_user(Map.new(params))
```

Comment:
If we're going to support this here, we should do it elsewhere when applicable.

Reply:
I think it is a good idea to allow keyword lists or maps for these request params. We should support it. If it's fine with you, I'll be doing something similar to this for the rest of the API changes.

Reply:
Yep, I'm fine with this approach.